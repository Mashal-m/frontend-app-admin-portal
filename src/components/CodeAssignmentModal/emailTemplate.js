const emailTemplate = `Your learning manager has provided you with a new access code to take a course at edX. You may redeem this code for {REDEMPTIONS_REMAINING} # of courses.

edX Login: {USER_EMAIL}
Access Code: {CODE}
Expiration Date: {EXPIRATION_DATE}

You can insert the access code at checkout under "coupon code" for applicable courses.

For any questions, please reach out to your Learning Manager.`;

export default emailTemplate;
