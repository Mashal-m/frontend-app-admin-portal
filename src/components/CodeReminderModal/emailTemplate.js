const emailTemplate = {
  greeting: ' This is a reminder that your Learning Manager has provided you with an access code to take a course at edX.',
  body: `You have redeemed this code {REDEEMED_OFFER_COUNT} time(s) out of {TOTAL_OFFER_COUNT} number of available course redemptions.

edX Login: {USER_EMAIL}
Access Code: {CODE}
Expiration Date: {EXPIRATION_DATE}`,
  closing: `You can insert the access code at checkout under "coupon code" for applicable courses.

For any questions, please reach out to your Learning Manager.`,
};

export default emailTemplate;
