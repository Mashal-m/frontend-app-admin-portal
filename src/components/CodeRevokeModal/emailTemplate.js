const emailTemplate = {
  body: 'Your learning manager has revoked {CODE} and it is no longer assigned to your edX account {USER_EMAIL}.',
  closing: `
  Please reach out to your Learning Manager for additional questions.
  `,
};

export default emailTemplate;
