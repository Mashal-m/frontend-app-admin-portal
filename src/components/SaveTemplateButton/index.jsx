import React from 'react';
import PropTypes from 'prop-types';
import { Button, Icon } from '@edx/paragon';
import { SubmissionError } from 'redux-form';

class SaveTemplateButton extends React.Component {
  constructor(props) {
    super(props);
    this.handleSaveTemplate = this.handleSaveTemplate.bind(this);
  }

  handleSaveTemplate() {
    const {
      setMode,
      templateData,
      saveTemplate,
      templateType,
    } = this.props;

    const options = {
      email_type: templateType,
      email_greeting: templateData['email-template-greeting'],
      email_closing: templateData['email-template-closing'],
    };

    setMode('save');

    return saveTemplate(options)
      .catch((error) => {
        // backend to frontend field name map
        const fieldMap = {
          email_greeting: 'email-template-greeting',
          email_closing: 'email-template-closing',
        };
        const { response, message } = error;
        const errors = {
          _error: [],
        };

        Object.entries(fieldMap).forEach(([key, value]) => {
          if (response && response.data && response.data[key]) {
            errors[value] = response.data[key];
          }
        });
        /* eslint-disable no-underscore-dangle */
        // when no field specific error returned then show the main exception message
        if (Object.keys(errors).length === 1 && Object.keys(errors)[0] === '_error') {
          errors._error.push(message);
        }
        /* eslint-enable no-underscore-dangle */

        throw new SubmissionError(errors);
      });
  }

  render() {
    const {
      saving,
      disabled,
      buttonLabel,
      handleSubmit,
    } = this.props;
    const saveButtonIconClasses = saving ? ['fa-spinner', 'fa-spin'] : [];

    return (
      <Button
        className="btn-primary save-template-btn"
        disabled={disabled || saving}
        onClick={handleSubmit(this.handleSaveTemplate)}
      >
        <React.Fragment>
          <Icon className={`fa mr-2 ${saveButtonIconClasses.join(' ')}`} />
          {buttonLabel}
        </React.Fragment>
      </Button>
    );
  }
}

SaveTemplateButton.defaultProps = {
  saving: false,
  disabled: true,
  buttonLabel: 'Save Template',
};

SaveTemplateButton.propTypes = {
  templateType: PropTypes.string.isRequired,
  templateData: PropTypes.shape({
    'email-template-greeting': PropTypes.string.isRequired,
    'email-template-closing': PropTypes.string.isRequired,
  }).isRequired,
  saveTemplate: PropTypes.func.isRequired,
  setMode: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  saving: PropTypes.bool,
  disabled: PropTypes.bool,
  buttonLabel: PropTypes.string,
};

export default SaveTemplateButton;
