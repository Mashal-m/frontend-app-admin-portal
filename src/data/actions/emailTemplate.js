import {
  EMAIL_TEMPLATE_REQUEST,
  EMAIL_TEMPLATE_SUCCESS,
  EMAIL_TEMPLATE_FAILURE,
} from '../constants/emailTemplate';

import EcommerceApiService from '../services/EcommerceApiService';

const emailTemplateRequest = () => ({
  type: EMAIL_TEMPLATE_REQUEST,
});

const emailTemplateSuccess = data => ({
  type: EMAIL_TEMPLATE_SUCCESS,
  payload: {
    data,
  },
});

const emailTemplateFailure = error => ({
  type: EMAIL_TEMPLATE_FAILURE,
  payload: {
    error,
  },
});

const fetchEmailTemplates = options => (
  (dispatch) => {
    dispatch(emailTemplateRequest());
    return EcommerceApiService.fetchEmailTemplate(options)
      .then((response) => {
        dispatch(emailTemplateSuccess(response.data));
      })
      .catch((error) => {
        dispatch(emailTemplateFailure(error));
      });
  }
);

export default fetchEmailTemplates;
