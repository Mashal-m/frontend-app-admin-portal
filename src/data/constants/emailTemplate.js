const EMAIL_TEMPLATE_REQUEST = 'EMAIL_TEMPLATE_REQUEST';
const EMAIL_TEMPLATE_SUCCESS = 'EMAIL_TEMPLATE_SUCCESS';
const EMAIL_TEMPLATE_FAILURE = 'EMAIL_TEMPLATE_FAILURE';
const SAVE_TEMPLATE_REQUEST = 'SAVE_TEMPLATE_REQUEST';
const SAVE_TEMPLATE_SUCCESS = 'SAVE_TEMPLATE_SUCCESS';
const SAVE_TEMPLATE_FAILURE = 'SAVE_TEMPLATE_FAILURE';


export {
  EMAIL_TEMPLATE_REQUEST,
  EMAIL_TEMPLATE_SUCCESS,
  EMAIL_TEMPLATE_FAILURE,
  SAVE_TEMPLATE_REQUEST,
  SAVE_TEMPLATE_SUCCESS,
  SAVE_TEMPLATE_FAILURE,
};
