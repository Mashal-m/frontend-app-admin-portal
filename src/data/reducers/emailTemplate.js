import {
  EMAIL_TEMPLATE_SUCCESS,
  EMAIL_TEMPLATE_REQUEST,
  EMAIL_TEMPLATE_FAILURE,
} from '../constants/emailTemplate';

import { transformTemplates } from '../../utils';
import assignEmailTemplate from '../../components/CodeAssignmentModal/emailTemplate';
import remindEmailTemplate from '../../components/CodeReminderModal/emailTemplate';
import revokeEmailTemplate from '../../components/CodeRevokeModal/emailTemplate';

const initialState = {
  loading: false,
  error: null,
  assign: {
    'email-template-greeting': assignEmailTemplate.greeting,
    'email-template-body': assignEmailTemplate.body,
    'email-template-closing': assignEmailTemplate.closing,
  },
  remind: {
    'email-template-greeting': remindEmailTemplate.greeting,
    'email-template-body': remindEmailTemplate.body,
    'email-template-closing': remindEmailTemplate.closing,
  },
  revoke: {
    'email-template-greeting': revokeEmailTemplate.greeting,
    'email-template-body': revokeEmailTemplate.body,
    'email-template-closing': revokeEmailTemplate.closing,
  },
};

const emailTemplate = (state = initialState, action) => {
  switch (action.type) {
    case EMAIL_TEMPLATE_REQUEST:
      return {
        loading: true,
        error: null,
      };
    case EMAIL_TEMPLATE_SUCCESS:
      return {
        loading: false,
        error: null,
        ...transformTemplates(action.payload.data, initialState),
      };
    case EMAIL_TEMPLATE_FAILURE:
      return {
        loading: false,
        error: action.payload.error,
      };
    default:
      return state;
  }
};

export default emailTemplate;
