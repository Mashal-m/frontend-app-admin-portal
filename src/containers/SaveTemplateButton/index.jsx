import { connect } from 'react-redux';

import { saveTemplate } from '../../data/actions/emailTemplate';
import SaveTemplateButton from '../../components/SaveTemplateButton';

const mapStateToProps = state => ({
  saving: state.emailTemplate.saving,
});

const mapDispatchToProps = dispatch => ({
  saveTemplate: options => new Promise((resolve, reject) => {
    dispatch(saveTemplate({
      options,
      onSuccess: (response) => { resolve(response); },
      onError: (error) => { reject(error); },
    }));
  }),
});

export default connect(mapStateToProps, mapDispatchToProps)(SaveTemplateButton);
