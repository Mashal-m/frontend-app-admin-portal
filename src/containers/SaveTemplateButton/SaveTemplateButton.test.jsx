import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import EcommerceApiService from '../../data/services/EcommerceApiService';
import { SAVE_TEMPLATE_REQUEST } from '../../data/constants/emailTemplate';
import SaveTemplateButton from './index';

jest.mock('../../data/services/EcommerceApiService');
const mockStore = configureMockStore([thunk]);
const initialState = {
  portalConfiguration: {
    enterpriseId: 'test-enterprise-id',
  },
  emailTemplate: {
    saving: false,
  },
};
const store = mockStore({
  ...initialState,
});
const saveTemplateData = {
  email_type: 'assign',
  email_greeting: 'Greeting',
  email_closing: 'Closing',
};
const templateType = saveTemplateData.email_type;
const saveTemplateSpy = jest.spyOn(EcommerceApiService, 'saveTemplate');

const SaveTemplateButtonWrapper = props => (
  <MemoryRouter>
    <Provider store={store}>
      <SaveTemplateButton
        templateType={templateType}
        templateData={{
          'email-template-greeting': saveTemplateData.email_greeting,
          'email-template-closing': saveTemplateData.email_closing,
        }}
        setMode={() => {}}
        handleSubmit={submitFunction => submitFunction}
        {...props}
      />
    </Provider>
  </MemoryRouter>
);

describe('<SaveTemplateButton />', () => {
  it('renders correctly in disabled state', () => {
    const tree = renderer
      .create((
        <SaveTemplateButtonWrapper />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly in enabled state', () => {
    const tree = renderer
      .create((
        <SaveTemplateButtonWrapper disabled={false} />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly while saving a template', () => {
    const newStore = mockStore({
      emailTemplate: {
        saving: true,
      },
    });

    const tree = renderer
      .create((
        <SaveTemplateButtonWrapper store={newStore} />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls saveTemplate on click with correct data', () => {
    const expectedActions = [{
      type: SAVE_TEMPLATE_REQUEST,
      payload: { emailType: templateType },
    }];
    const successResponse = {
      email_greeting: saveTemplateData.email_greeting,
      email_body: 'email body',
      email_closing: saveTemplateData.email_closing,
    };
    EcommerceApiService.saveTemplate.mockImplementation(() => Promise.resolve({
      data: successResponse,
    }));
    const wrapper = mount((
      <SaveTemplateButtonWrapper disabled={false} />
    ));

    wrapper.find('button').find('.btn.btn-primary').simulate('click');
    expect(store.getActions()).toEqual(expectedActions);
    expect(saveTemplateSpy).toHaveBeenCalledWith(saveTemplateData);
  });
});
