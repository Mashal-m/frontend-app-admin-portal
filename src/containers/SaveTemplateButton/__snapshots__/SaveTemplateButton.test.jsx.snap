// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SaveTemplateButton /> renders correctly in disabled state 1`] = `
<button
  className="btn btn-primary save-template-btn"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  type="button"
>
  <span
    aria-hidden={true}
    className="fa mr-2 "
    id="Icon1"
  />
  Save Template
</button>
`;

exports[`<SaveTemplateButton /> renders correctly in enabled state 1`] = `
<button
  className="btn btn-primary save-template-btn"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  type="button"
>
  <span
    aria-hidden={true}
    className="fa mr-2 "
    id="Icon2"
  />
  Save Template
</button>
`;

exports[`<SaveTemplateButton /> renders correctly while saving a template 1`] = `
<button
  className="btn btn-primary save-template-btn"
  disabled={true}
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  type="button"
>
  <span
    aria-hidden={true}
    className="fa mr-2 fa-spinner fa-spin"
    id="Icon3"
  />
  Save Template
</button>
`;
